FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py

```diff
@@ -0,0 +1,289 @@
+    df["MUY"] = df_ref.muy
+    df["PLUSY"] = propagate_with_plus(dphi=df_ref.muy, init=y)
+    df["MINUSY"] = propagate_with_minus(dphi=df_ref.muy, init=y)
+    print(df_ref[["mux", "muy"]])
+
+    for idx in range(n):
+        with_error = MadXBoundaryConditions(
+            alfx=get_realization(x.alpha),
+            betx=get_realization(x.beta),
+            alfy=get_realization(y.alpha),
+            bety=get_realization(y.beta),
+        )
+
+        df_err = generate_df(madx, sequence="fodo", boundary_conditions=with_error)
+        
+        diff_x = (df_err.mux - df_ref.mux) % 1
+        diff_y = (df_err.muy - df_ref.muy) % 1
+        diff_x[diff_x > 0.5] -= 1
+        diff_y[diff_y > 0.5] -= 1
+
+        df[f"MADXX{idx:03d}"] = diff_x
+        df[f"MADXY{idx:03d}"] = diff_y
+
+    tfs.write(output, df, save_index="NAME")
+    return df
+
+
+# Helper ---
+
+def generate_fodo(sequence: str ="fodo") -> Madx:
+    """Generate a standard FODO sequence.
+    
+    Args:
+        sequence (str): Sequence name.
+    """
+    madx = Madx()
+    madx.beam()
+    madx.globals["mqf.k1"] = 0.3037241107
+    madx.globals["mqd.k1"] = -madx.globals["mqf.k1"]
+    madx.input(f"""
+    {sequence}: sequence, l=10, refer=entry;
+    mqf: quadrupole, at=0, l=1, k1:=mqf.k1;
+    dff: drift,      at=1, l=4;
+    mqd: quadrupole, at=5, l=1, k1:=mqd.k1;
+    dfd: drift,      at=6, l=4;
+    endsequence;
+    """)
+    madx.use(sequence=sequence)
+    return madx
+
+
+def generate_df(madx: Madx, sequence: str, boundary_conditions: MadXBoundaryConditions = None) -> pd.DataFrame:
+    """Generate a dataframe with Mad-X twiss and the given boundary conditions.
+    
+    Args:
+        madx (Madx): Mad-X cpymad object.
+        sequence (str): Sequence name.
+        boundary_conditions (MadXBoundaryConditions): Boundary conditions.
+    """
+    if boundary_conditions is None:
+        boundary_conditions = MadXBoundaryConditions()
+
+    madx.use(sequence=sequence)
+    madx.twiss(sequence=sequence, chrom=True, **boundary_conditions.as_dict())
+    return madx.table.twiss.dframe().copy()
+
+
+
+# Propagate Errors -------------------------------------------------------------
+
+def propagate_with_minus(dphi: FloatOrArray, init: PropagableBoundaryConditions):
+    """Phase error propagation as given in Eq. (2) of the paper,
+    but the first plus sign replaced with a minus sign.
+
+    Args:
+        dphi (FloatOrArray): Phase advance(s). 
+        init (PropagableBoundaryConditions): Initial conditions.
+
+    Returns:
+        FloatOrArray: Phase error at the given phase advance(s).
+    """
+    alpha, erralpha = init.as_tuple(init.alpha)
+    beta, errbeta = init.as_tuple(init.beta)
+
+    res = np.sqrt(
+        (
+                (((np.cos(4 * np.pi * dphi) - 1) * alpha) - 
+                 (np.sin(4 * np.pi * dphi))
+                 ) * 0.5 * errbeta/beta
+        ) ** 2 +
+        ((np.cos(4*np.pi*dphi) - 1) * 0.5 * erralpha) ** 2
+    ) / (2*np.pi)
```

Comment:
making this formula a fct with input of sign avoids copy-paste errors and maybe better readable

Reply:
if you have the inputs as ufloats, can't you also use the direct formula phi= ... and let uncertainties do the math to then get the err on phi afterwards?